src/python/pants/util/meta.py

```diff
@@ -138,3 +139,41 @@ def new_setattr(self, key: str, value: Any) -> None:
   cls.__init__ = new_init  # type: ignore
   cls.__setattr__ = new_setattr  # type: ignore
   return cls
+
+
+class EnumMatchError(ValueError):
+  """Issue when using match() on an enum."""
+
+
+class InexhaustiveMatchError(EnumMatchError):
+  """Not all values of the enum specified in the pattern match."""
+
+
+class UnrecognizedMatchError(EnumMatchError):
+  """A value is used that is not a part of the enum."""
+
+
+_E = TypeVar('_E', bound='Enum')
+_V = TypeVar('_V')
+
+
+def match(enum_instance: Enum, enum_values_to_results: Mapping[_E, _V]) -> _V:
+  # TODO: consider memoizing the result of this entire method, as well as the value of
+  # `all_instances` for a given enum class!
+  all_instances = frozenset(type(enum_instance).__members__.values())
+  unrecognized_values = [
+    value for value in enum_values_to_results if value not in all_instances
+  ]
+  missing_values = [
+    value for value in all_instances if value not in enum_values_to_results
+  ]
+  if unrecognized_values:
+    raise UnrecognizedMatchError(
+      f"Match includes values not defined in the enum. Unrecognized: {unrecognized_values}"
+    )
+  if missing_values:
+    raise InexhaustiveMatchError(
+      f"All enum values must be covered by the match. Missing: {missing_values}"
+    )
+  typed_enum = cast(_E, enum_instance)
```

Comment:
Why isn't `enum_instance` simply typed as `_E` from the get go?

Reply:
I think this was leftover from when we originally used `self` here when it was a method. We should try removing this line and see if MyPy still works.